src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -41,10 +45,94 @@ public RouteEndpointDataSource(IServiceProvider applicationServices, bool throwO
+        {
+            var requestDelegate = (RequestDelegate)handler;
+
+            // Use request delegate directly when there are no filters.
+            if (options.EndpointBuilder is null || options.EndpointBuilder.FilterFactories.Count == 0)
+            {
+                return requestDelegate;
+            }
+
+            return BuildFilterPipeline(requestDelegate, options);
+        }
+    }
+
+    private static RequestDelegate BuildFilterPipeline(RequestDelegate requestDelegate, RequestDelegateFactoryOptions options)
+    {
+        Debug.Assert(options.EndpointBuilder != null);
+
+        var serviceProvider = options.ServiceProvider ?? options.EndpointBuilder.ApplicationServices;
+        var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? new JsonOptions();
+        var jsonSerializerOptions = jsonOptions.SerializerOptions;
+
+        var factoryContext = new EndpointFilterFactoryContext
+        {
+            MethodInfo = requestDelegate.Method,
+            ApplicationServices = options.EndpointBuilder.ApplicationServices
+        };
+        var jsonTypeInfo = (JsonTypeInfo<object>)jsonSerializerOptions.GetTypeInfo(typeof(object));
+
+        EndpointFilterDelegate filteredInvocation = async (EndpointFilterInvocationContext context) =>
+        {
+            if (context.HttpContext.Response.StatusCode >= 400)
+            {
+                return EmptyHttpResult.Instance;
+            }
+            else
+            {
+                await requestDelegate(context.HttpContext);
+                return EmptyHttpResult.Instance;
+            }
+        };
+
+        var initialFilteredInvocation = filteredInvocation;
+        for (var i = options.EndpointBuilder.FilterFactories.Count - 1; i >= 0; i--)
+        {
+            var currentFilterFactory = options.EndpointBuilder.FilterFactories[i];
+            filteredInvocation = currentFilterFactory(factoryContext, filteredInvocation);
+        }
+
+        // The filter factories have run without modifying per-request behavior, we can skip running the pipeline.
+        if (ReferenceEquals(initialFilteredInvocation, filteredInvocation))
+        {
+            return requestDelegate;
+        }
+
+        return async (HttpContext httpContext) =>
+        {
+            var obj = await filteredInvocation(new DefaultEndpointFilterInvocationContext(httpContext, new object[] { httpContext }));
+            if (obj is not null)
+            {
+                await ExecuteHandlerHelper.ExecuteReturnAsync(obj, httpContext, jsonSerializerOptions, jsonTypeInfo);
+            }
+        };
+    }
+
+    // Due to cyclic references between Http.Extensions and
+    // Http.Results, we define our own instance of the `EmptyHttpResult`
+    // type here.
+    private sealed class EmptyHttpResult : IResult
```

Comment:
This is no good, we need to fix this (it's what [@captainsafia's change here is about](https://github.com/dotnet/aspnetcore/pull/45878))